traits/trait_list_object.py

```diff
@@ -59,34 +59,117 @@ def __repr__(self):
 
 
 def _normalize_index(index, length):
-    """ Normalize integer index to range 0 to len (inclusive). """
+    """ Normalize integer index to range 0 to len (inclusive).
+
+    These rules reflect the normalization that Python uses for the
+    ``insert`` and ``pop`` operations.
+
+    """
+
     index = operator.index(index)
     if index < 0:
         return max(0, length + index)
     else:
         return min(length, index)
 
 
-def _normalize_slice(index, length):
-    """ Normalize slice start and stop to range 0 to len (inclusive). """
+def _normalize_slice_or_index(index, length):
+    """ Normalize a slice or index for use with __delitem__ or __setitem__.
 
-    # Non-extended slice (step = 1): return only need the start index.
-    if index.step is None or index.step == 1:
-        if index.start is None:
-            return 0
-        else:
-            return _normalize_index(index.start, length)
+    For slices with positive step, returns a slice that's equivalent for the
+    purposes of __delitem__ and __setitem__ operations. For slices with
+    negative step, a normalized slice representing the reverse of the given
+    slice is returned: note that in this case, the matching *added* and
+    *removed* lists will need to be reversed.
+
+    Slices with a step of 1 or -1 are normalized to a single integer index,
+    referring to the position of the first element referenced by the slice.
+
+    Similarly, slices that refer to only a single element of the corresponding
+    list (for example, a slice of `[1::10]` applied to a list of length 5)
+    are normalized to the index that refers to that same element.
+
+    Empty slices are also normalized to a single index. Note that in the case
+    of an empty slice, the corresponding __delitem__ or __setitem__ operation
+    does not cause any list change, so does not issue a notification. So the
+    normalized index in this case is unused in current code.
+
+    A normalized slice will have 0 <= start < stop <= length and a step >= 2.
+    It should further satisfy start + step < stop. The stop will always be
+    one larger than the last element referenced by the slice.
 
-    # Extended slice with negative step: do not normalize.
-    if index.step is not None and index.step < 0:
-        return index
+    A normalized index will satisfy 0 <= index <= length.
```

Comment:
Oops, I overlooked this. If index is negative enough, `index + length` may still be negative, so it may not be larger or equal to 0.

Reply:
True, but in that case the corresponding operation will fail, so we'll never use that index. IOW, there's no change to the logic needed here, but there may be a comment or docstring change needed.

Reply:
Updating the docstring sounds good to me!

Reply:
Done.